Document Service Provider usage (#20)
servicebinding/website: service-provider.md

permalink: /service-provider/
---

Service providers expose bindings through a `Secret` resource with data required for connectivity.  The specification's [Provisioned Service section][provisioned-service] describes this in detail.  Alternatively, the specification also supports [Direct Secret Reference][direct-secret-reference].  The only requirement for Direct Secret Reference is a `Secret` resource with data required for connectivity.  Alternatively, if you are creating a Provisioned Service (preferred approach), you also need a custom resource with `.status.binding.name` attribute pointing to the `Secret` resource.

Here is an example custom resource:

```yaml
apiVersion: example.dev/v1beta1
kind: Database
metadata:
  name: database-service
...
status:
  binding:
    name: production-db-secret
```

> {% include note.html content="Implementing the Provisioned Service contract frees the `ServiceBinding` from needing to know about the name of the `Secret` holding the credentials.  The service can update the secret name exposed over time.  This behavior not only decouples the ServiceBinding and the workload from needing to know the name of the Secret, it also enables use of immutable secrets.  An immutable secret cannot be updated after it is created.  To rotate the credentials, a new secret can be created and update the same in the Provisioned Service resource's `.status.binding.name` attribute." %}

Reviewer: nit: The ServiceBinding needs to know the name of the secret, but the author of that resource doesn't

```suggestion
{% include note.html content="Implementing the Provisioned Service contract frees the `ServiceBinding` creator from needing to know about the name of the `Secret` holding the credentials.  The service can update the secret name exposed over time.  This behavior not only decouples the ServiceBinding and the workload from needing to know the name of the Secret, it also enables use of immutable secrets.  An immutable secret cannot be updated after it is created.  To rotate the credentials, a new secret can be created and update the same in the Provisioned Service resource's `.status.binding.name` attribute." %}
```

In the previous example, `production-db-secret` is the name of the `Secret` resource with data entries required for connectivity.  The `Secret` resource should contain a `type` entry that can be used for identifying the service.  It helps the application to identify the service as a relational database, key-value store, or a cache server.  There is no standardization on the value for `type`, but you can see some good examples in the [Spring Cloud Bindings][spring-cloud-bindings].  A few examples:

- `postgresql`
- `mysql`
- `redis`
- `mongodb`

> The `Secret` resource's [`type` field](https://kubernetes.io/docs/concepts/configuration/secret/#secret-types) should have a special value based on the `type` data entry value.  The expected format is `servicebinding.io/{type}`.  For example, if the value for `type` data entry is `mysql`, then the `Secret` resource's `type` field value should be `servicebinding.io/mysql`.  This recommendation helps to query `Secret` resources of particular type using field-selector. For example:

Reviewer: it's worth adding a caveat that the type is encouraged but not mandatory, and has no runtime effect. If the secret has a different type that provides significant meaning, that value should be used instead.

Reviewer 2: @baijum did this get addressed?

| `private-key` | A PEM-encoded private key used in mTLS client authentication

If there is any entry that doesn’t follow the given requirement, you can choose different names. For example, if there is a URI-like string but not a valid one, as per RFC-3986, use another name (e.g., "custom-uri"). If a service has multiple values for a key, consider a dot-delimited prefix (e.g. "amqp.port" or "mqtt.port") reusing the prefix with other keys as appropriate.

# Considerations for Role-Based Access Control (RBAC)

As a service provider, you can create a `ClusterRole` with the label
`servicebinding.io/controller=true` and the verbs `get`, `list`, and `watch`
listed in the rules.  Service Binding implementations use these permissions to lookup the service resource and read the binding secret name from its status when referenced by a ServiceBinding.  Here is an example `ClusterRole`:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: examlples-service-bindings
  labels:
    servicebinding.io/controller: "true" # matches the aggregation rule selector
rules:
- apiGroups:

>   - example.dev

Reviewer 2: @baijum can we get a response to this please?

permalink: /service-provider/
---

Service providers expose bindings through a Secret resource with data required for connectivity.  The specification's [Provisioned Service section][provisioned-service] describes this in detail.  But there are two important parts:

1. A Secret resource with data required for connectivity

> 2. A custom resource with `.status.binding.name` attribute pointing to the Secret resource

Reviewer 3: I think we should specify that (2) is optional, and considered best practice if a CR is involved, as you could directly point to a Secret from a `ServiceBinding` CR.

Author: I have updated the text based on this feedback.

> Service providers expose bindings through a Secret resource with data required for connectivity.  The specification's [Provisioned Service section][provisioned-service] describes this in detail.  Alternatively, the specification also support [Direct Secret Reference][direct-secret-reference].  The only important part required for Direct Secret Reference is a Secret resource with data required for connectivity.  But if you are creating a Provisioned Service (preferred approach), you also need a custom resource with `.status.binding.name` attribute pointing to the Secret resource.

Reviewer 2: ```suggestion
Service providers expose bindings through a `Secret` resource with data required for connectivity.  The specification's [Provisioned Service section][provisioned-service] describes this in detail.  Alternatively, the specification also support [Direct Secret Reference][direct-secret-reference].  The only important part required for Direct Secret Reference is a `Secret` resource with data required for connectivity.  But if you are creating a Provisioned Service (preferred approach), you also need a custom resource with `.status.binding.name` attribute pointing to the `Secret` resource.
```

> ```

Reviewer 2: ```suggestion
```yaml
```

> apiVersion: exmple.dev/v1beta1

Reviewer 2: ```suggestion
apiVersion: example.dev/v1beta1
```

kind: Database
metadata:
  name: database-service
...
status:
  binding:
    name: production-db-secret
```

> In the above example, `production-db-secret` is the name of the Secret resource with data entries required for connectivity.  The Secret resource should contain a `type` entry that can be used for identifying the service.  It helps the application to indentify the service as a relational database, key-value store, or a cache server.  There is no standardization on the value for type, but you can see some good examples in the Spring Cloud Bindings. Few examples:

Reviewer 2: ```suggestion
In the above example, `production-db-secret` is the name of the `Secret` resource with data entries required for connectivity.  The `Secret` resource should contain a `type` entry that can be used for identifying the service.  It helps the application to indentify the service as a relational database, key-value store, or a cache server.  There is no standardization on the value for `type`, but you can see some good examples in the Spring Cloud Bindings.  A few examples:
```

> In the above example, `production-db-secret` is the name of the Secret resource with data entries required for connectivity.  The Secret resource should contain a `type` entry that can be used for identifying the service.  It helps the application to indentify the service as a relational database, key-value store, or a cache server.  There is no standardization on the value for type, but you can see some good examples in the Spring Cloud Bindings. Few examples:

Reviewer 2: *suggestion*: Perhaps a link to the Spring Cloud Bindings documentation?

- cassandra

> - couchbase

Reviewer 2: ```suggestion
- `couchbase`
```

- db2

> - elasticsearch

Reviewer 2: ```suggestion
- `elasticsearch`
```

> - kafka

Reviewer 2: ```suggestion
- `kafka`
```

- ldap

> - mongodb

Reviewer 2: ```suggestion
- `mongodb`
```

> - mysql

Reviewer 2: ```suggestion
- `mysql`
```

> - neo4j

Reviewer 2: ```suggestion
- `neo4j`
```

> - oracle

Reviewer 2: ```suggestion
- `oracle`
```

> - postgresql

Reviewer 2: ```suggestion
- `postgresql`
```

> - rabbitmq

Reviewer 2: ```suggestion
- `rabbitmq`
```

> - redis

Reviewer 2: ```suggestion
- `redis`
```

> - sqlserver

Reviewer 2: ```suggestion
- `sqlserver`
```

> - vault

Reviewer 2: ```suggestion
- `vault`
```

> The type of the Secret resource should have a special value based on the `type` entry value.  The format is `servicebinding.io/{type}`.  For example, if the type entry is `mysql` the Secret's type value should be `servicebinding.io/mysql`.  This recommendation helps to query Secret resources of particular type using field-selector. For example:

Reviewer 2: *thought*: It feels like the use of `type` here is really confusing because, depending on where you are in a given sentence, you either mean a type key in the payload or the type key in the spec.  It might be worthwhile being a bit more explicit about which you mean in every usage.

Author: I tried to address this issue by mentioning "field" when referring to the type of resource. And use "data" entry value for data value.
fe5638987aedd460d44e3d6e339fc7fc0faf25ce

> ```

Reviewer 2: ```suggestion
```bash
```

kubectl get secrets --field-selector="type=servicebinding.io/type"
```

> will give the Secret resources of `mysql` type.

Reviewer 2: ```suggestion
will give the `Secret` resources of `mysql` type.
```

Similar to `type` entry, spec also recommends to add a `provider` entry to identify the provider.  The `provider` entry is a further classification of the type.

> Here is an example Secret:

Reviewer 2: ```suggestion
Here is an example `Secret`:
```

> ```

Reviewer 2: ```suggestion
```yaml
```

apiVersion: v1
kind: Secret
metadata:
  name: production-db-secret
type: servicebinding.io/mysql
stringData:
  type: mysql
  provider: bitnami
  host: localhost
  port: 3306
  username: root
  password: root
```

# Well-known Secret Entries

Apart from the special `type`, and `provider` entries in the Secret data, there are few special words, if used must follow certain restrictions for the values.  These are called well-known entries.

| Name | Requirements
| ---- | ------------
| `host` | A DNS-resolvable host name or IP address
| `port` | A valid port number
| `uri` | A valid URI as defined by [RFC3986](https://tools.ietf.org/html/rfc3986)
| `username` | A string-based username credential
| `password` | A string-based password credential
| `certificates` | A collection of PEM-encoded X.509 certificates, representing a certificate chain used in mTLS client authentication
| `private-key` | A PEM-encoded private key used in mTLS client authentication

> If there is any entry that doesn’t follow the given requirement, you can choose different names. For example, if there is a URI-like string but not a valid one, as per RFC-3986, use another name (e.g., "custom-uri").

Reviewer 2: *praise*: Great explicit callout about what to do here!

# Considerations for Role-Based Access Control (RBAC)

> As a service povider, you can create a `ClusterRole` with the label

Reviewer 2: ```suggestion
As a service provider, you can create a `ClusterRole` with the label
```

> (`servicebinding.io/controller=true`) and the verbs (`get`, `list`, and `watch`)

Reviewer 2: ```suggestion
`servicebinding.io/controller=true` and the verbs `get`, `list`, and `watch`
```

permalink: /service-provider/
---

> Service providers expose bindings through a `Secret` resource with data required for connectivity.  The specification's [Provisioned Service section][provisioned-service] describes this in detail.  Alternatively, the specification also support [Direct Secret Reference][direct-secret-reference].  The only important part required for Direct Secret Reference is a `Secret` resource with data required for connectivity.  But if you are creating a Provisioned Service (preferred approach), you also need a custom resource with `.status.binding.name` attribute pointing to the `Secret` resource.

Reviewer 3: ```suggestion
Service providers expose bindings through a `Secret` resource with data required for connectivity.  The specification's [Provisioned Service section][provisioned-service] describes this in detail.  Alternatively, the specification also supports [Direct Secret Reference][direct-secret-reference].  The only requirement for Direct Secret Reference is a `Secret` resource with data required for connectivity.  Alternatively, if you are creating a Provisioned Service (preferred approach), you also need a custom resource with `.status.binding.name` attribute pointing to the `Secret` resource.
```

> Service providers expose bindings through a `Secret` resource with data required for connectivity.  The specification's [Provisioned Service section][provisioned-service] describes this in detail.  Alternatively, the specification also support [Direct Secret Reference][direct-secret-reference].  The only important part required for Direct Secret Reference is a `Secret` resource with data required for connectivity.  But if you are creating a Provisioned Service (preferred approach), you also need a custom resource with `.status.binding.name` attribute pointing to the `Secret` resource.

Author: Added a slightly modified text. Please check.
